### PR TITLE
Add gradient for operation 'SparseSlice' 

### LINKED ISSUE #19373: No gradient defined for operation 'SparseSlice'
i am working on solve a indeterminate equation optimization with sparse_slice functions and met this error.
it's strange all sparse functions have gradient except this one. any suggestion how can i avoid this problem?

### Source code / logs
import tensorflow as tf
sparse_values = tf.identity(tf.Variable(tf.constant([1., 2., 3.])))
sparse_indices = tf.constant([[0, 0], [1, 1], [2, 2]], dtype=tf.int64)
sparse_matrix = tf.SparseTensor(sparse_indices, sparse_values, [3, 3])
sparse_slice = tf.sparse_slice(sparse_matrix,[0,0], [2,2])
multiplied = tf.sparse_tensor_dense_matmul(sparse_slice, tf.eye(2))
loss = tf.reduce_sum(multiplied)
gradients = tf.gradients(loss, [sparse_values])
with tf.Session() as session:
    tf.global_variables_initializer().run()
    print(session.run(gradients))

### PR DESCRIPTION
**Pull request by facaiy**
Fix #19373.

### CODE REVIEW

File: tensorflow/core/kernels/sparse_slice_grad_op.cc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    Tensor *val_grad;
+    OP_REQUIRES_OK(ctx,
+                   ctx->allocate_output(0, TensorShape({input_nnz}), &val_grad));
+
+    T *val_grad_flat = val_grad->flat<T>().data();
+    const T *backprop_val_grad_flat = backprop_val_grad->flat<T>().data();
+    memset(val_grad_flat, 0, sizeof(T) * input_nnz);
+
+    // Fill gradients for position where indices of input and ouput are same.
```
Comment (rachellim) [inline tensorflow/core/kernels/sparse_slice_grad_op.cc]: nit: output

File: tensorflow/core/kernels/sparse_slice_grad_op.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        }
+      }
+      if (is_same) {
+        val_grad_flat[i] = backprop_val_grad_flat[j];
+        ++j;
+      }
+    }
+    OP_REQUIRES(
+        ctx, backprop_val_grad->NumElements() == j,
+        errors::Internal("Elements of backprop_val_grad aren't eaten up :",
```
Comment (rachellim) [inline tensorflow/core/kernels/sparse_slice_grad_op.cc]: Suggestion: ```"Elements of backprop_val_grad aren't all propagated. Num elements: ", backprop_val_grad->NumElements(), ", used: ", j```

File: tensorflow/core/kernels/sparse_slice_grad_op.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    OP_REQUIRES(
+        ctx, backprop_val_grad->NumElements() == output_indices->dim_size(0),
+        errors::InvalidArgument("# elements of backprop_val_grad and # rows of "
+                                "output_indices should match (#nnz of sum): got ",
+                                backprop_val_grad->NumElements(), " and ",
+                                output_indices->dim_size(0)));
+    OP_REQUIRES(ctx, TensorShapeUtils::IsVector(input_start->shape()),
+                errors::InvalidArgument(
+                    "Input start should be a vector but received shape ",
+                    input_start->shape().DebugString()));
```
Comment (rachellim) [inline tensorflow/core/kernels/sparse_slice_grad_op.cc]: Should validate the size of the `input_start` vector too

**Review by rachellim — COMMENTED** [review]
Thanks for the PR! This looks good, some minor comments

**Comment by facaiy** [comment]
@rachellim Thank you for your review. I think all your comments have been resolved, could you take a look again?

**Review by rachellim — DISMISSED** [review]
Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rachellim** [comment]
Hey @facaiy -- apologies about the delay pushing this PR through! Looks like you have some lint failures  [(see the log)](https://source.cloud.google.com/results/invocations/06ee8145-56ad-4acb-9eea-4ac830829874/log), primarily to do with indentation. Can you fix them?

**Review by rachellim — APPROVED** [review]

**Comment by facaiy** [comment]
@rachellim Thank you. I have fixed them.
